lib/contextual_logger.rb

```diff
@@ -43,31 +44,40 @@ def normalize_message(message)
   end
 
   module LoggerMixin
+    include Context
+
     delegate :register_secret, :register_secret_regex, to: :redactor
 
+    def global_context
+      @global_context ||= {}.freeze
+    end
+
     def global_context=(context)
-      Context::Handler.new(context).set!
+      @global_context = context.freeze
+    end
+
+    def current_context
+      current_context_override || global_context
     end
 
+    # TODO: Deprecate current_context_for_thread in v2.0.
+    alias current_context_for_thread current_context
+
     def with_context(context)
-      context_handler = Context::Handler.new(current_context_for_thread.deep_merge(context))
-      context_handler.set!
+      previous_context_override = current_context_override
+      self.current_context_override = current_context.deep_merge(context)
```

Comment:
❓ Should we be freezing the merged context to assure ourselves context mutation cannot be done?

Reply:
Sounds fine, but it would need a `deep_freeze` wouldn't it? It doesn't look like that's built in. I could write it inline, I guess?
```
class << self
  def deep_freeze(object)
    case object
    when Hash
      object.values.each { deep_freeze(_1) }
    when Array
      object.each { deep_freeze(_1) }
    end
    object.freeze
  end
end
```
Seem worth it?

Reply:
Hmm, now I realize the my `global_context.freeze` has the same issue!

Reply:
We agreed to live with plain `.freeze`. That will catch most cases without require a special inline `deep_freeze`.